daemon.go

```diff
@@ -2,215 +2,124 @@ package main
-	if err := d.Queue.Create(); err != nil {
-		return err
+// NewDaemon creates a new Daemon.
+func NewDaemon(instanceID string, handler Handler, listeners ...Listener) *Daemon {
+	return &Daemon{
+		instanceID: instanceID,
+		handler:    handler,
+		listeners:  listeners,
 	}
+}
 
-	// ensure the queue deletion happens only once
-	var deleteOnce sync.Once
-	defer func() {
-		deleteOnce.Do(d.deleteQueue)
-	}()
-
-	if err := d.Queue.Subscribe(); err != nil {
-		return err
-	}
-	defer func() {
-		if err := d.Queue.Unsubscribe(); err != nil {
-			log.WithError(err).Error("Failed to unsubscribe from sns topic")
-		}
-	}()
+// Start the Daemon.
+func (d *Daemon) Start(ctx context.Context) error {
+	// Add a child context to stop all listeners when one has returned
+	listenerCtx, stopListening := context.WithCancel(ctx)
+	defer stopListening()
 
-	ch := make(chan *sqs.Message)
+	// Use a buffered channel to avoid deadlocking a goroutine when we stop listening
+	notices := make(chan TerminationNotice, 1)
 
-	go func() {
-		for m := range ch {
-			var env Envelope
-			var msg AutoscalingMessage
-
-			// unmarshal outer layer
-			if err := json.Unmarshal([]byte(*m.Body), &env); err != nil {
-				log.WithError(err).Info("Failed to unmarshal envelope")
-				continue
-			}
+	var wg sync.WaitGroup
+	for _, listener := range d.listeners {
+		wg.Add(1)
 
-			log.WithFields(log.Fields{
-				"type":    env.Type,
-				"subject": env.Subject,
-			}).Debugf("Received an SQS message")
+		l := log.WithField("listener", listener.Type())
 
-			// unmarshal inner layer
-			if err := json.Unmarshal([]byte(env.Message), &msg); err != nil {
-				log.WithError(err).Info("Failed to unmarshal autoscaling message")
-				continue
-			}
+		go func() {
+			defer wg.Done()
+			defer stopListening()
 
-			if msg.InstanceID != d.InstanceID {
-				log.WithFields(log.Fields{
-					"was":    msg.InstanceID,
-					"wanted": d.InstanceID,
-				}).Debugf("Skipping autoscaling event, doesn't match instance id")
-				continue
+			if err := listener.Start(listenerCtx, notices); err != nil {
+				l.WithError(err).Error("Failed to start listener")
+			} else {
+				l.Info("Stopped listener")
 			}
+		}()
+		l.Info("Starting listener")
+	}
 
-			d.handleMessage(ctx, msg, func() {
-				// delete the queue before we complete
-				deleteOnce.Do(d.deleteQueue)
-			})
-		}
-	}()
-
-	spotTerminations := pollSpotTermination(ctx)
 	go func() {
-		for notice := range spotTerminations {
-			log.Infof("Got a spot instance termination notice: %v", notice)
-
-			log.Info("Executing handler")
-			timer := time.Now()
-			err := executeHandler(ctx, d.Handler, []string{terminationTransition, d.InstanceID})
-			executeCtx := log.WithFields(log.Fields{
-				"duration": time.Now().Sub(timer),
-			})
-
-			if err != nil {
-				executeCtx.WithError(err).Error("Handler script failed")
-				return
-			}
-
-			executeCtx.Info("Handler finished successfully")
-		}
+		wg.Wait()
+		close(notices)
```

Comment:
Now that I think about it, the solution I've implemented here _does not_ guarantee that a termination notice will be read off the channel for the following reason:
































When a notice is received and sent over the channel, the listener immediately exits and calls [stopListening()](https://github.com/buildkite/lifecycled/blob/d027ee63319490b6ffef531cf13736dbf51d7531/daemon.go#L47) which will then stop all other listeners. This means that there is actually a race between the goroutine waiting to close the channel, and the main thread which is trying to read (just one notice) from the channel before it is closed.